{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieve RD FASTA sequences\n",
    "#### DATE: 07-08-19\n",
    "#### TASK: \n",
    "####       - Retrive coordinates of RD from the RD.csv file\n",
    "####       - Use coords to extracts RD in FASTA from the H37rv reference genome\n",
    "####       - Create blast db from RDs\n",
    "####       - Blast mbovis/morygis against local db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, io, random, subprocess\n",
    "import string\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width',150)\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio import AlignIO, SeqIO\n",
    "\n",
    "sys.path.append('/home/tortoise/pybioviz/')\n",
    "from pybioviz import viewers, utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The H37rv genome is going to be converted to a string, to be sliced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we open up the H37rv genome.\n",
    "# The RD are defined by coordinates on this genome so we will use the RD sequences spliced from this\n",
    "def extract_genome(genomefile):\n",
    "    \n",
    "    # This string will have each line of genomic sequence added to it\n",
    "    genome_string = ''\n",
    "    \n",
    "    # The genome file is opened here\n",
    "    with open(genomefile) as gfile:\n",
    "    \n",
    "        # Each line is read\n",
    "        for line in gfile:\n",
    "    \n",
    "            # We skip the fasta header\n",
    "            if '>' in line:\n",
    "                continue\n",
    "            # The genome string is then assembled...\n",
    "            else:\n",
    "                genome_string = genome_string + line\n",
    "    # and returned    \n",
    "    return genome_string\n",
    "\n",
    "# The retrieval of the genome is called\n",
    "MTBenome = extract_genome('ref_genomes/MTB_h37rv.fna')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The .csv file is going to be opened, with relevent data extracted and stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function reads the csv and takes needed information\n",
    "# It stores the uses an inputted dictionary\n",
    "# The key added is each RD name\n",
    "# The value added is the slice of the H37rv genome using the csv Start and Stop coordinates as slice coordinates\n",
    "# RDict is the empty dictionary used to store info\n",
    "def extract_csv(csv_file, RDict, RDcount):\n",
    "    \n",
    "    # RD is the dataframe made from the .csv file\n",
    "    # Commented .csv inputs are ignored\n",
    "    RD = pd.read_csv(csv_file)#, comment='#')\n",
    "    \n",
    "    # Each row is iterated over using iterrows\n",
    "    # This creates a set of each row, where each value can be called using the column name\n",
    "    for iteration, row in RD.iterrows():\n",
    "        \n",
    "        # RDict takes in the row's RD name as a key..\n",
    "        # .. and a genome slice as a value\n",
    "        # The slice is specified using the RD's Start and Stop coordinates\n",
    "        RDict[row.RD_name] = MTBenome[row.Start:row.Stop+1]\n",
    "        mbo_RDcount[row.RD_name] = 0\n",
    "        mor_RDcount[row.RD_name] = 0\n",
    "\n",
    "# This empty dict later cathes .csv information of each RD\n",
    "RDict = {}\n",
    "\n",
    "# These dict is going to be used to count results for each RD\n",
    "mbo_RDcount = {}\n",
    "mor_RDcount = {}\n",
    "\n",
    "# The .csv file is parsed through our extraction function\n",
    "extract_csv('RD.csv', RDict, RDcount)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here the extracted files are written in fasta format to an output file for later blasting\n",
    "# A new file is created\n",
    "with open('RD_seq.fasta', 'w+') as fna:\n",
    "    \n",
    "    # We iterate over the RD dictionary's items\n",
    "    for RD, seq in RDict.items():\n",
    "        \n",
    "        # And write into the output file each RD and its sequence..\n",
    "        # .. in fasta format\n",
    "        # FASTA format is needed for blasting\n",
    "        fna.write(f'>{RD}\\n{seq}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we create a database of the RDs, which can align each genome against\n",
    "# The function takes in the created RD fasta file\n",
    "def makedb(RDfna):\n",
    "    \n",
    "    # a command is then created using the inputted file\n",
    "    # The inputted file is the database to be created, using the -in parameter\n",
    "    # -dtype is nucl as we are creating a nucleotide database\n",
    "    # The database will then be called the same as the input file\n",
    "    make_db = f'makeblastdb -in {RDfna} -dbtype nucl'\n",
    "    \n",
    "    # The command is used on the command line\n",
    "    os.system(make_db)\n",
    "    \n",
    "    # A return could be used here to pass on the name of the database created..\n",
    "    \n",
    "# The database is created using our created file    \n",
    "makedb('RD_seq.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -query ref_genomes/mbovis.fna -db RD_seq.fasta -out mbovisblast.txt -outfmt 7\n",
      "blastn -query ref_genomes/morygis_ncbi.fna -db RD_seq.fasta -out morygis_ncbiblast.txt -outfmt 7\n"
     ]
    }
   ],
   "source": [
    "# Here we perform the blast search with our db\n",
    "def do_blast(genome):\n",
    "    \n",
    "    # This is the created output file name, made frpom whichever genome is being parsed\n",
    "    outfile = os.path.basename(genome).split('.')[0] + 'blast.txt'\n",
    "    \n",
    "    # The cmd is then:\n",
    "    \n",
    "    # - created\n",
    "    blastcmd = f'blastn -query {genome} -db RD_seq.fasta -out {outfile} -outfmt 7'\n",
    "    \n",
    "    # - printed\n",
    "    print(blastcmd)\n",
    "    \n",
    "    # - called on the cmd line\n",
    "    os.system(blastcmd)\n",
    "    \n",
    "    # the output file name is returned, and stored as a variable\n",
    "    return outfile\n",
    "\n",
    "# the blastin function is called for each genome, and the name of the output file is stored\n",
    "mbo_blast = do_blast('ref_genomes/mbovis.fna')\n",
    "mor_blast = do_blast('ref_genomes/morygis_ncbi.fna')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mbovis hit count: 65\n",
      "morygis hit count: 78\n"
     ]
    }
   ],
   "source": [
    "## We then read the blast results, taking in the blast file\n",
    "def read_blast_res(blast_res, RDcount):\n",
    "    # This is just a simple count of how many hits total came up\n",
    "    RD_hit_count = 0\n",
    "    \n",
    "    # The results file is opened\n",
    "    with open(blast_res) as res:\n",
    "        \n",
    "        # Each line is read\n",
    "        for line in res:\n",
    "            \n",
    "            #ignoring commented lines..\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            \n",
    "            # ignoring low identity hits..\n",
    "            elif float(line.split('\\t')[2]) <= 70 or float(line.split('\\t')[3]) <= 200:\n",
    "                continue\n",
    "            \n",
    "            \n",
    "            # Each result is printed/counted\n",
    "            else:\n",
    "                #print(line)\n",
    "                RD_hit_count += 1\n",
    "                RDcount[line.split('\\t')[1]] += 1\n",
    "    \n",
    "    # The hit count is then returned\n",
    "    return RD_hit_count\n",
    "\n",
    "# Each blast count is called..\n",
    "mbo_RD_hits = read_blast_res(mbo_blast, mbo_RDcount)\n",
    "mor_RD_hits = read_blast_res(mor_blast, mor_RDcount)\n",
    "\n",
    "# ..and hit count is printed\n",
    "print(f'mbovis hit count: {mbo_RD_hits}')\n",
    "print(f'morygis hit count: {mor_RD_hits}')\n",
    "\n",
    "## Beware that jupyter lab does not reset each dictionary, hence results are appended. Reset in previous tab."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'#RD9': 2, 'RD711': 1, 'RD702': 1, '#RD4': 1, '#RD1bcg': 4, '#RD1mic': 3, '#RD2seal': 1, '#RD2bcg': 2, '#RD7': 1, '#RD8': 1, '#RD10': 1, '#RD12bov': 1, '#RD12can': 1, 'RD105': 1, 'RD239': 1, 'RD750': 1, 'RD142': 1, 'RD150': 2, 'RD181': 1, 'RD207': 1, 'RD115': 1, 'RD122': 1, 'RD174': 0, 'RD182': 2, 'RD183': 1, 'RD193': 1, 'RD219': 1, 'RD724': 1, 'RD726': 1, 'RD761': 1, '7bp pks15/1': 0, 'RD11': 3, 'RD1': 4, 'RD2': 2, 'RD3': 2, 'RD4': 1, 'RD5': 1, 'RD6': 3, 'RD7': 1, 'RD8': 1, 'RD9': 4, 'RD10': 1, 'RD12': 1, 'RD13': 1, 'RD14': 2}\n",
      "{'#RD9': 5, 'RD711': 1, 'RD702': 1, '#RD4': 1, '#RD1bcg': 4, '#RD1mic': 3, '#RD2seal': 1, '#RD2bcg': 4, '#RD7': 1, '#RD8': 1, '#RD10': 1, '#RD12bov': 1, '#RD12can': 1, 'RD105': 1, 'RD239': 1, 'RD750': 1, 'RD142': 1, 'RD150': 2, 'RD181': 1, 'RD207': 1, 'RD115': 1, 'RD122': 1, 'RD174': 0, 'RD182': 2, 'RD183': 1, 'RD193': 1, 'RD219': 1, 'RD724': 1, 'RD726': 1, 'RD761': 1, '7bp pks15/1': 0, 'RD11': 3, 'RD1': 4, 'RD2': 4, 'RD3': 2, 'RD4': 1, 'RD5': 1, 'RD6': 4, 'RD7': 1, 'RD8': 1, 'RD9': 8, 'RD10': 1, 'RD12': 1, 'RD13': 1, 'RD14': 3}\n"
     ]
    }
   ],
   "source": [
    "print(mbo_RDcount)\n",
    "print(mor_RDcount)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mbovis</th>\n",
       "      <th>morygis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>#RD10</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD12bov</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD12can</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD1bcg</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD1mic</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD2bcg</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD2seal</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD4</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD7</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD8</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#RD9</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7bp pks15/1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD1</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD10</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD105</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD11</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD115</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD12</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD122</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD13</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD14</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD142</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD150</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD174</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD181</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD182</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD183</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD193</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD2</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD207</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD219</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD239</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD3</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD4</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD5</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD6</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD7</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD702</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD711</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD724</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD726</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD750</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD761</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD8</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RD9</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mbovis  morygis\n",
       "#RD10             1        1\n",
       "#RD12bov          1        1\n",
       "#RD12can          1        1\n",
       "#RD1bcg           4        4\n",
       "#RD1mic           3        3\n",
       "#RD2bcg           2        4\n",
       "#RD2seal          1        1\n",
       "#RD4              1        1\n",
       "#RD7              1        1\n",
       "#RD8              1        1\n",
       "#RD9              2        5\n",
       "7bp pks15/1       0        0\n",
       "RD1               4        4\n",
       "RD10              1        1\n",
       "RD105             1        1\n",
       "RD11              3        3\n",
       "RD115             1        1\n",
       "RD12              1        1\n",
       "RD122             1        1\n",
       "RD13              1        1\n",
       "RD14              2        3\n",
       "RD142             1        1\n",
       "RD150             2        2\n",
       "RD174             0        0\n",
       "RD181             1        1\n",
       "RD182             2        2\n",
       "RD183             1        1\n",
       "RD193             1        1\n",
       "RD2               2        4\n",
       "RD207             1        1\n",
       "RD219             1        1\n",
       "RD239             1        1\n",
       "RD3               2        2\n",
       "RD4               1        1\n",
       "RD5               1        1\n",
       "RD6               3        4\n",
       "RD7               1        1\n",
       "RD702             1        1\n",
       "RD711             1        1\n",
       "RD724             1        1\n",
       "RD726             1        1\n",
       "RD750             1        1\n",
       "RD761             1        1\n",
       "RD8               1        1\n",
       "RD9               4        8"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RD_hits_df = pd.DataFrame([mbo_RDcount, mor_RDcount])\n",
    "RD_hits_df.index=['mbovis', 'morygis']\n",
    "RD_hits_df.T.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AC'"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MTBenome[10:12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
