{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## this notebook is a test of nucdiff using some fake sequence data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, io, random, subprocess\n",
    "import string\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width',600)\n",
    "from Bio.Seq import Seq, MutableSeq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio import AlignIO, SeqIO\n",
    "\n",
    "sys.path.append('/home/tortoise/pybioviz/')\n",
    "from pybioviz import utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make two test sequence and make gaps and some mutations in the second"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13387\n",
      "8734\n",
      "26014\n",
      "5678\n",
      "18888\n",
      "5132\n",
      "30347\n",
      "12320\n",
      "29034\n",
      "32696\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec = SeqIO.parse('ref_genomes/mbovis.fna',format='fasta').__next__()\n",
    "\n",
    "s1=rec.seq[:50000] \n",
    "seq1 = SeqRecord(s1,id='test1',description='my ref')\n",
    "#create altered seq\n",
    "\n",
    "s2 = (s1[:200]+ 10*'ATATATATATAGCGCGCGATGCTAG' +s1[200:400]+s1[400:500]+s1[600:18000]+s1[20000:30000]+s1[35000:40000]).tomutable()\n",
    "s2[200] = 'T'\n",
    "s2[1000] = 'A'\n",
    "\n",
    "bases = ['A', 'T', 'G', 'C']\n",
    "for i in range(10):\n",
    "    position = random.randrange(len(s2))\n",
    "    base = bases[random.randrange(len(bases))]\n",
    "    while s2[position] == base:\n",
    "        position = random.randrange(len(s2))\n",
    "    print(position)\n",
    "    s2[position] = base\n",
    "\n",
    "\n",
    "seq2 = SeqRecord(s2,id='test2',description='my query')\n",
    "\n",
    "SeqIO.write(seq1,'test1.fna','fasta')\n",
    "SeqIO.write(seq2,'test2.fna','fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucmer --maxgap=500 --mincluster=100 --coords -p nucmer test1.fna test2.fna\n"
     ]
    }
   ],
   "source": [
    "nuc = utils.align_nucmer('test1.fna','test2.fna')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S1</th>\n",
       "      <th>E1</th>\n",
       "      <th>S2</th>\n",
       "      <th>E2</th>\n",
       "      <th>LEN 1</th>\n",
       "      <th>LEN 2</th>\n",
       "      <th>IDENT</th>\n",
       "      <th>TAG1</th>\n",
       "      <th>TAG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>500</td>\n",
       "      <td>1</td>\n",
       "      <td>500</td>\n",
       "      <td>500</td>\n",
       "      <td>500</td>\n",
       "      <td>99.80</td>\n",
       "      <td>test1</td>\n",
       "      <td>test2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>601</td>\n",
       "      <td>18000</td>\n",
       "      <td>501</td>\n",
       "      <td>17900</td>\n",
       "      <td>17400</td>\n",
       "      <td>17400</td>\n",
       "      <td>99.99</td>\n",
       "      <td>test1</td>\n",
       "      <td>test2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20001</td>\n",
       "      <td>30001</td>\n",
       "      <td>17901</td>\n",
       "      <td>27901</td>\n",
       "      <td>10001</td>\n",
       "      <td>10001</td>\n",
       "      <td>100.00</td>\n",
       "      <td>test1</td>\n",
       "      <td>test2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>35000</td>\n",
       "      <td>40000</td>\n",
       "      <td>27900</td>\n",
       "      <td>32900</td>\n",
       "      <td>5001</td>\n",
       "      <td>5001</td>\n",
       "      <td>100.00</td>\n",
       "      <td>test1</td>\n",
       "      <td>test2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      S1     E1     S2     E2  LEN 1  LEN 2   IDENT   TAG1   TAG2\n",
       "0      1    500      1    500    500    500   99.80  test1  test2\n",
       "1    601  18000    501  17900  17400  17400   99.99  test1  test2\n",
       "2  20001  30001  17901  27901  10001  10001  100.00  test1  test2\n",
       "3  35000  40000  27900  32900   5001   5001  100.00  test1  test2"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### run nucdiff and find the differences!\n",
    "\n",
    "* run nucdiff on the 2 samples with test1 as ref and find deletions plus snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>del_len</th>\n",
       "      <th>ins_len</th>\n",
       "      <th>query_bases</th>\n",
       "      <th>ref_bases</th>\n",
       "      <th>query_coord</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SV_1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>248</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>201-448</td>\n",
       "      <td>199</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SV_3</td>\n",
       "      <td>deletion</td>\n",
       "      <td>100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>750</td>\n",
       "      <td>500</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SV_4</td>\n",
       "      <td>deletion</td>\n",
       "      <td>2000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18150</td>\n",
       "      <td>18000</td>\n",
       "      <td>20000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SV_5</td>\n",
       "      <td>deletion</td>\n",
       "      <td>4998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28151</td>\n",
       "      <td>30001</td>\n",
       "      <td>34999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP_1</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5133-5133</td>\n",
       "      <td>4982</td>\n",
       "      <td>4983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP_2</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>5679-5679</td>\n",
       "      <td>5528</td>\n",
       "      <td>5529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP_3</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>8735-8735</td>\n",
       "      <td>8584</td>\n",
       "      <td>8585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP_4</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>12321-12321</td>\n",
       "      <td>12170</td>\n",
       "      <td>12171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNP_5</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>13388-13388</td>\n",
       "      <td>13237</td>\n",
       "      <td>13238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SNP_6</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>18889-18889</td>\n",
       "      <td>20738</td>\n",
       "      <td>20739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SNP_7</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>26015-26015</td>\n",
       "      <td>27864</td>\n",
       "      <td>27865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SNP_8</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>29035-29035</td>\n",
       "      <td>35884</td>\n",
       "      <td>35885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SNP_9</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>30348-30348</td>\n",
       "      <td>37197</td>\n",
       "      <td>37198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SNP_10</td>\n",
       "      <td>substitution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>32697-32697</td>\n",
       "      <td>39546</td>\n",
       "      <td>39547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID          Name del_len ins_len query_bases ref_bases  query_coord  start    end\n",
       "0    SV_1     insertion     NaN     248         NaN       NaN      201-448    199    200\n",
       "2    SV_3      deletion     100     NaN         NaN       NaN          750    500    600\n",
       "3    SV_4      deletion    2000     NaN         NaN       NaN        18150  18000  20000\n",
       "4    SV_5      deletion    4998     NaN         NaN       NaN        28151  30001  34999\n",
       "0   SNP_1  substitution     NaN     NaN           A         G    5133-5133   4982   4983\n",
       "1   SNP_2  substitution     NaN     NaN           C         A    5679-5679   5528   5529\n",
       "2   SNP_3  substitution     NaN     NaN           G         C    8735-8735   8584   8585\n",
       "3   SNP_4  substitution     NaN     NaN           T         C  12321-12321  12170  12171\n",
       "4   SNP_5  substitution     NaN     NaN           G         A  13388-13388  13237  13238\n",
       "5   SNP_6  substitution     NaN     NaN           A         G  18889-18889  20738  20739\n",
       "6   SNP_7  substitution     NaN     NaN           G         T  26015-26015  27864  27865\n",
       "7   SNP_8  substitution     NaN     NaN           T         C  29035-29035  35884  35885\n",
       "8   SNP_9  substitution     NaN     NaN           A         T  30348-30348  37197  37198\n",
       "9  SNP_10  substitution     NaN     NaN           C         G  32697-32697  39546  39547"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featurekeys = ['ID', 'Name', 'del_len', 'ins_len', 'query_bases', 'ref_bases', 'query_coord', 'start', 'end']\n",
    "\n",
    "def features_to_dataframe(features, cds=False):\n",
    "    \"\"\"Get features from a biopython seq record object into a dataframe\n",
    "    Args:\n",
    "        features: bio seqfeatures\n",
    "       returns: a dataframe with a row for each cds/entry.\n",
    "      \"\"\"\n",
    "\n",
    "    #preprocess features\n",
    "    allfeat = []\n",
    "    for (item, f) in enumerate(features):\n",
    "        x = f.__dict__\n",
    "        q = f.qualifiers\n",
    "        #print(q)\n",
    "        x.update(q)\n",
    "        d = {}\n",
    "        d['start'] = f.location.start\n",
    "        d['end'] = f.location.end\n",
    "        #d['strand'] = f.location.strand\n",
    "        for i in featurekeys:\n",
    "            if i in x:\n",
    "                if type(x[i]) is list:\n",
    "                    d[i] = x[i][0]\n",
    "                else:\n",
    "                    d[i] = x[i]\n",
    "        allfeat.append(d)\n",
    "\n",
    "    df = pd.DataFrame(allfeat,columns=featurekeys)\n",
    "\n",
    "    return df\n",
    "\n",
    "feats = utils.gff_to_features('Nucdiff_test12/results/first_ref_struct.gff')\n",
    "feat_structs = features_to_dataframe(feats)\n",
    "deletions = feat_structs[(feat_structs.Name=='deletion') | (feat_structs.Name=='insertion')]\n",
    "\n",
    "\n",
    "feats_snps = utils.gff_to_features('Nucdiff_test12/results/first_ref_snps.gff')\n",
    "feat_snps = features_to_dataframe(feats_snps)\n",
    "snps = feat_snps[feat_snps.Name=='substitution']\n",
    "\n",
    "\n",
    "snpdels = pd.concat([deletions, snps])\n",
    "snpdels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
